### PR TITLE
Refactor `MostSpecificHostMatch`

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

To prepare fix for #38082

### CODE REVIEW

File: pilot/pkg/model/config.go [hunk 1]
```diff
@@ -279,28 +279,67 @@ func resolveGatewayName(gwname string, meta config.Meta) string {
...
 
 	// exact match first
 	if m != nil {
 		if _, ok := m[needle]; ok {
 			return needle, true
 		}
+	}
+
+	if needle.IsWildCarded() {
+		// slice has better loop performance than map, so use stack to range
```
Comment (howardjohn) [inline pilot/pkg/model/config.go:295]: nit: comment is stale

**Review by howardjohn — COMMENTED** [review]
/test benchmark_istio

**Comment by howardjohn** [comment]
I think this might have wiped out some of the improvements in https://github.com/istio/istio/pull/28427

```
MostSpecificHostMatch/10Exact-8                         21.4ns ± 1%        20.9ns ± 0%    -1.92%  (p=0.008 n=5+5)
MostSpecificHostMatch/50Exact-8                         25.6ns ± 0%        21.9ns ± 0%   -14.54%  (p=0.008 n=5+5)
MostSpecificHostMatch/100Exact-8                        21.2ns ± 0%        22.0ns ± 1%    +3.43%  (p=0.008 n=5+5)
MostSpecificHostMatch/1000Exact-8                       20.0ns ± 0%        22.3ns ± 1%   +11.46%  (p=0.008 n=5+5)
MostSpecificHostMatch/5000Exact-8                       23.7ns ± 0%        21.6ns ± 0%    -8.86%  (p=0.008 n=5+5)
MostSpecificHostMatch/10DestRuleWildcard-8              95.5ns ± 0%       267.0ns ± 1%  +179.43%  (p=0.016 n=4+5)
MostSpecificHostMatch/50DestRuleWildcard-8               215ns ± 1%         920ns ± 1%  +327.85%  (p=0.008 n=5+5)
MostSpecificHostMatch/100DestRuleWildcard-8              817ns ± 0%        2303ns ± 1%  +182.01%  (p=0.008 n=5+5)
MostSpecificHostMatch/1000DestRuleWildcard-8            7.97µs ± 0%       24.64µs ± 0%  +209.25%  (p=0.008 n=5+5)
MostSpecificHostMatch/5000DestRuleWildcard-8            20.1µs ± 0%       105.7µs ± 0%  +426.65%  (p=0.016 n=5+4)
MostSpecificHostMatch/10NeedleWildcard-8                99.4ns ± 0%       295.7ns ± 2%  +197.61%  (p=0.008 n=5+5)
MostSpecificHostMatch/50NeedleWildcard-8                 428ns ± 0%        1202ns ± 1%  +180.73%  (p=0.008 n=5+5)
MostSpecificHostMatch/100NeedleWildcard-8                835ns ± 1%        2470ns ± 1%  +195.77%  (p=0.008 n=5+5)
MostSpecificHostMatch/1000NeedleWildcard-8              8.01µs ± 1%       24.76µs ± 0%  +209.09%  (p=0.008 n=5+5)
MostSpecificHostMatch/5000NeedleWildcard-8              12.5µs ± 0%       100.8µs ± 1%  +704.52%  (p=0.016 n=4+5)
```

I guess iterate over an array is a lot faster than over map...

But in the real code, unlike the benchmark, we pay the price of constructing+sorting the list, so possible worth it. WDYT?

**Review by howardjohn — APPROVED** [review]

**Comment by hzxuzhonghu** [comment]
```
                                                                            added sort                      with this patch
BenchmarkMostSpecificHostMatch/10DestRuleWildcard-8       2688355	       443.6 ns/op    4557777	       263.4 ns/op
BenchmarkMostSpecificHostMatch/50DestRuleWildcard-8        288608	      4147 ns/op      1313739	       913.3 ns/op
BenchmarkMostSpecificHostMatch/100DestRuleWildcard-8       105112	     11394 ns/op       531706	      2254 ns/op
BenchmarkMostSpecificHostMatch/1000DestRuleWildcard-8        6057	    197232 ns/op        50526	     23671 ns/op
BenchmarkMostSpecificHostMatch/5000DestRuleWildcard-8        1015	   1179864 ns/op        10000	    103228 ns/op
BenchmarkMostSpecificHostMatch/10NeedleWildcard-8         2723353	       438.9 ns/op    3934935	       304.7 ns/op
BenchmarkMostSpecificHostMatch/50NeedleWildcard-8          275935	      4345 ns/op       941943	      1280 ns/op
BenchmarkMostSpecificHostMatch/100NeedleWildcard-8         105061	     11405 ns/op       465379	      2565 ns/op
BenchmarkMostSpecificHostMatch/1000NeedleWildcard-8          6061	    197064 ns/op        46158	     26251 ns/op
BenchmarkMostSpecificHostMatch/5000NeedleWildcard-8          1022	   1172023 ns/op        10000	    101863 ns/op
```

@howardjohn

**Comment by howardjohn** [comment]
Ok so this one is faster since we do not sort, and the benchmark did not include sort. LGTM. feel free to remove the hold

**Pull request merged**